File: Queue/DoubleEndedQueue/DoubleEndedQueue.py
class Node:
    def __init__(self, value):
        self.data = value
        self.next = None
        self.prev = None

class Deque:
    def __init__(self):
        self.front = None
        self.rear = None

    def is_empty(self) -> bool:
        return self.front is None

    def add_to_front(self, element):
        new_node = Node(element)
        if self.is_empty():
            self.front = self.rear = new_node
        else:
            new_node.next = self.front
            self.front.prev = new_node
            self.front = new_node

    def add_to_rear(self, element):
        new_node = Node(element)
        if self.is_empty():
            self.front = self.rear = new_node
        else:
            new_node.prev = self.rear
            self.rear.next = new_node
            self.rear = new_node

    def remove_from_front(self) -> int:
        if self.is_empty():
            raise RuntimeError("Deque is Empty")
        value = self.front.data
        temp = self.front
        if self.front == self.rear:
            self.front = self.rear = None
        else:
            self.front = self.front.next
            self.front.prev = None
        del temp
        return value

    def remove_from_rear(self) -> int:
        if self.is_empty():
            raise RuntimeError("Deque is Empty")
        value = self.rear.data
        temp = self.rear
        if self.front == self.rear:
            self.front = self.rear = None
        else:
            self.rear = self.rear.prev
            self.rear.next = None
        del temp
        return value

    def size(self) -> int:
        count = 0
        current = self.front
        while current is not None:
            current = current.next
            count += 1
        return count

    def print_deque(self):
        current = self.front
        print("NULL", end="")
        while current is not None:
            print(" <->", current.data, end="")
            current = current.next
        print(" <-> NULL")

if __name__ == "__main__":
    deque = Deque()

    while True:
        print("\n1. Insert at Front\n2. Insert at End\n3. Delete from Front\n4. Delete from End\n5. Print\n6. Size\n7. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            data = int(input("Enter data: "))
            deque.add_to_front(data)
        elif choice == 2:
            data = int(input("Enter data: "))
            deque.add_to_rear(data)
        elif choice == 3:
            try:
                print("Removed from front:", deque.remove_from_front())
            except RuntimeError as e:
                print("Error:", e)
        elif choice == 4:
            try:
                print("Removed from end:", deque.remove_from_rear())
            except RuntimeError as e:
                print("Error:", e)
        elif choice == 5:
            print("Deque elements:", end=" ")
            deque.print_deque()
        elif choice == 6:
            print("Deque size:", deque.size())
        elif choice == 7:
            print("Exiting the program.")
            break
        else:
            print("Invalid choice. Please enter a valid option.")

File: Queue/DoubleEndedQueue/DoubleEndedQueue.cpp
// Using Linked List

#include <iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;
    Node* prev;
    Node(int value) : data(value), next(nullptr), prev(nullptr) {}
};

class Deque {
private:
    Node* front;
    Node* rear;
public:
    Deque() : front(nullptr), rear(nullptr) {}

    bool isEmpty() const {
        return front == nullptr;
    }

    void addToFront(int element) {
        Node* newNode = new Node(element);
        if (isEmpty()) {
            front = rear = newNode;
        } else {
            newNode->next = front;
            front->prev = newNode;
            front = newNode;
        }
    }

    void addToRear(int element) {
        Node* newNode = new Node(element);
        if (isEmpty()) {
            front = rear = newNode;
        } else {
            newNode->prev = rear;
            rear->next = newNode;
            rear = newNode;
        }
    }

    int removeFromFront() {
        if (isEmpty()) {
            cout<<"Deque is Empty";
            throw runtime_error(".");
        }
        int rearValue = front->data;
        Node* temp = front;
        if (front == rear) {
            front = rear = nullptr;
        } else {
            front = front->next;
            front->prev = nullptr;
        }
        delete temp;
        return rearValue;
    }

    int removeFromRear() {
        if (isEmpty()) {
            cout<<"Deque is Empty";
            throw runtime_error(".");
        }
        int rearValue = rear->data;
        Node* temp = rear;
        if (front == rear) {
            front = rear = nullptr;
        } else {
            rear = rear->prev;
            rear->next = nullptr;
        }
        delete temp;
        return rearValue;
    }

    int size() const {
        int count = 0;
        Node* current = front;
        while (current != nullptr) {
            current = current->next;
            count++;
        }
        return count;
    }

    void printDeque() const {
        Node* current = front;
        cout<<"NULL";
        while (current != nullptr) {
            cout<<" <-> "<<current->data;
            current = current->next;
        }
        cout<<" <-> "<<"NULL";
    }

    ~Deque() {
        while (front != nullptr) {
            Node* temp = front;
            front = front->next;
            delete temp;
        }
    }
};

int main() {
    int ch = 0, data = 0;
    Deque deque;

    do {
        cout << "\n1. Insert at Front\n2. Insert at End\n3. Delete from Front\n4. Delete from End\n5. Print\n6. Size\n7. Exit\nEnter your choice: ";
        cin >> ch;

        switch (ch) {
            case 1:
                cout << "Enter data: ";
                cin >> data;
                deque.addToFront(data);
                break;
            case 2:
                cout << "Enter data: ";
                cin >> data;
                deque.addToRear(data);
                break;
            case 3:
                try {
                    cout << "Removed from front: " << deque.removeFromFront() << endl;
                } catch (const exception& e) {
                    cerr << e.what() << endl;
                }
                break;
            case 4:
                try {
                    cout << "Removed from end: " << deque.removeFromRear() << endl;
                } catch (const exception& e) {
                    cerr << e.what() << endl;
                }
                break;
            case 5:
                cout << "Deque elements: ";
                deque.printDeque();
                cout << endl;
                break;
            case 6:
                cout << "Deque size: " << deque.size() << endl;
                break;
            case 7:
                cout << "Exiting the program.\n";
                break;
            default:
                cout << "Invalid choice. Please enter a valid option.\n";
        }
    } while (ch != 7);

    return 0;
}